walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,124 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+
+	private static final int END_OF_DAY = 1440;
+
+	public Collection<TimeRange> query(Collection<Event> events,
+		MeetingRequest request) {
+		int duration = (int) request.getDuration();
+		Collection<String> attendees = request.getAttendees();
+		Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+		Collection<TimeRange> availableTimesWithOptionals = new ArrayList<>();
+
+		availableTimes = getAvaliableTimes(attendees, events, duration);
+
+		if (!optionalAttendees.isEmpty()) {
+			availableTimesWithOptionals = getAvaliableTimes(optionalAttendees,
+				events, duration);
+		}
+
+		if (optionalAttendees.isEmpty()) {
+			return availableTimes;
+		}
+
+		if (attendees.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		availableTimesWithOptionals = intersection(availableTimesWithOptionals,
+			availableTimes);
+
+		if (!availableTimesWithOptionals.isEmpty()) {
+			return availableTimesWithOptionals;
+		}
+
+		return availableTimes;
+	}
+
+	/**
+	 * Finds avaliable times for the given atteendee list, input 
+	 * collection and event duration
+	 */
+	private Collection<TimeRange> getAvaliableTimes(Collection<String> attendees,
+		Collection<Event> events, int duration) {
+		Collection<TimeRange> availableTimes = new ArrayList<>();
+
+		int start = 0;
+		int end = start + duration;
+		int availablePeriodStart = -1, availablePeriodEnd = -1;
+
+		while (end<= END_OF_DAY) {
+			TimeRange currentTimeRange = TimeRange.fromStartDuration(start, duration);
+			boolean works = true;
+
+			for (Event event: events) {
+				if (!works) break;
```

Comment:
Move this line to be the last line in this for loop. Then you can take out the `continue;` line on 84.